File: python/examples/symbols.py
#!/usr/bin/env python3
# Copyright (c) 2024-2026 Elias Bachaalany
# SPDX-License-Identifier: MPL-2.0
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at https://mozilla.org/MPL/2.0/.
#
# symbols: Symbol operations -- list, get, rename, restore, and delete symbols.
#
# Usage: python symbols.py [host_url]
#
# Defaults: http://127.0.0.1:18080, expects a program already open in Ghidra.

import sys

import libghidra as ghidra


def main() -> None:
    url = sys.argv[1] if len(sys.argv) >= 2 else "http://127.0.0.1:18080"
    client = ghidra.connect(url)

    # 1. List symbols (first page)
    try:
        sym_resp = client.list_symbols(limit=20)
    except ghidra.GhidraError as e:
        print(f"list_symbols failed: {e}", file=sys.stderr)
        sys.exit(1)

    symbols = sym_resp.symbols
    if not symbols:
        print("No symbols found -- is a program open?", file=sys.stderr)
        sys.exit(1)

    print(f"Symbols ({len(symbols)} shown):")
    for s in symbols:
        ns = f"  ns={s.namespace_name}" if s.namespace_name else ""
        flags = []
        if s.is_primary:
            flags.append("primary")
        if s.is_external:
            flags.append("external")
        if s.is_dynamic:
            flags.append("dynamic")
        flag_str = f"  [{', '.join(flags)}]" if flags else ""
        print(f"  0x{s.address:08x}  {s.name:<32}  type={s.type:<12}{ns}{flag_str}")

    # 2. Pick a function entry to work with
    try:
        func_resp = client.list_functions(limit=10)
    except ghidra.GhidraError as e:
        print(f"list_functions failed: {e}", file=sys.stderr)
        sys.exit(1)

    if not func_resp.functions:
        print("No functions found.", file=sys.stderr)
        sys.exit(1)

    target = func_resp.functions[0]
    addr = target.entry_address
    original_name = target.name
    print(f"\nTarget function: {original_name} at 0x{addr:08x}")

    # 3. Get the symbol at this address
    try:
        get_resp = client.get_symbol(addr)
    except ghidra.GhidraError as e:
        print(f"get_symbol failed: {e}", file=sys.stderr)
        sys.exit(1)

    if get_resp.symbol:
        s = get_resp.symbol
        print(f"  Symbol: id={s.symbol_id}  name={s.name}  full={s.full_name}  type={s.type}  source={s.source}")
    else:
        print("  No symbol at this address.")

    # 4. Rename the symbol
    new_name = "example_renamed_symbol"
    print(f"\nRenaming symbol at 0x{addr:08x} to '{new_name}'...")
    try:
        rename_resp = client.rename_symbol(addr, new_name)
        print(f"  renamed={rename_resp.renamed}  name={rename_resp.name}")
    except ghidra.GhidraError as e:
        print(f"  rename_symbol failed: {e}", file=sys.stderr)
        sys.exit(1)

    # Verify
    verify = client.get_symbol(addr)
    if verify.symbol:
        print(f"  Verified: symbol name is now '{verify.symbol.name}'")

    # 5. Restore the original name
    print(f"\nRestoring original name '{original_name}'...")
    try:
        restore_resp = client.rename_symbol(addr, original_name)
        print(f"  renamed={restore_resp.renamed}  name={restore_resp.name}")
    except ghidra.GhidraError as e:
        print(f"  rename_symbol failed: {e}", file=sys.stderr)

    # Verify restoration
    restored = client.get_symbol(addr)
    if restored.symbol and restored.symbol.name == original_name:
        print(f"  Verified: name restored to '{original_name}'")

    # 6. Show symbol deletion (on a secondary symbol if possible)
    #    We create a temporary rename, then delete it to demonstrate the API.
    temp_name = "temp_delete_me"
    print(f"\nDemonstrating delete_symbol...")
    print(f"  Renaming to '{temp_name}' first...")
    client.rename_symbol(addr, temp_name)

    try:
        del_resp = client.delete_symbol(addr, name_filter=temp_name)
        print(f"  deleted={del_resp.deleted}  deleted_count={del_resp.deleted_count}")
    except ghidra.GhidraError as e:
        print(f"  delete_symbol failed: {e}", file=sys.stderr)
        # Restore if delete didn't work
        client.rename_symbol(addr, original_name)

    # Final state
    final = client.get_symbol(addr)
    if final.symbol:
        print(f"  Final symbol: {final.symbol.name}")
    else:
        print("  Symbol removed.")

    print("\nDone.")


if __name__ == "__main__":
    main()

File: rust/examples/symbols.rs
// Copyright (c) 2024-2026 Elias Bachaalany
// SPDX-License-Identifier: MPL-2.0
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
// symbols: List symbols, inspect one, rename it, and restore the original name.
//
// Usage: symbols [host_url]

use libghidra as ghidra;

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let url = args.get(1).map_or("http://127.0.0.1:18080", |s| s.as_str());
    let client = ghidra::connect(url);

    // 1. List the first 20 symbols
    let syms = match client.list_symbols(0, u64::MAX, 20, 0) {
        Ok(s) => s,
        Err(e) => {
            eprintln!("ListSymbols failed (is a program open?): {e}");
            std::process::exit(1);
        }
    };

    println!("Symbols ({} shown):", syms.symbols.len());
    for s in &syms.symbols {
        println!(
            "  0x{:08x}  {:<30} type={:<12} ns={} primary={}",
            s.address, s.name, s.r#type, s.namespace_name, s.is_primary
        );
    }

    if syms.symbols.is_empty() {
        eprintln!("No symbols found.");
        std::process::exit(1);
    }

    // 2. Pick a symbol to inspect (prefer a function symbol)
    let target = syms
        .symbols
        .iter()
        .find(|s| s.r#type == "Function")
        .unwrap_or(&syms.symbols[0]);

    let addr = target.address;
    let original_name = target.name.clone();
    println!(
        "\nSelected symbol: '{}' at 0x{:08x} (type={})",
        original_name, addr, target.r#type
    );

    // 3. Get the symbol by address
    let detail = client.get_symbol(addr).unwrap_or_else(|e| {
        eprintln!("GetSymbol failed: {e}");
        std::process::exit(1);
    });

    if let Some(s) = &detail.symbol {
        println!("\nSymbol details:");
        println!("  ID:          {}", s.symbol_id);
        println!("  Name:        {}", s.name);
        println!("  Full name:   {}", s.full_name);
        println!("  Type:        {}", s.r#type);
        println!("  Namespace:   {}", s.namespace_name);
        println!("  Source:      {}", s.source);
        println!("  Primary:     {}", s.is_primary);
        println!("  External:    {}", s.is_external);
        println!("  Dynamic:     {}", s.is_dynamic);
    }

    // 4. Rename the symbol
    let new_name = format!("{}_renamed", original_name);
    println!("\nRenaming '{}' -> '{}'...", original_name, new_name);
    let rename = client.rename_symbol(addr, &new_name).unwrap_or_else(|e| {
        eprintln!("RenameSymbol failed: {e}");
        std::process::exit(1);
    });
    println!("  renamed={}, name=\"{}\"", rename.renamed, rename.name);

    // 5. Verify the rename
    let verify = client.get_symbol(addr).unwrap();
    if let Some(s) = &verify.symbol {
        println!("  Verified: symbol is now '{}'", s.name);
    }

    // 6. Restore original name
    println!("\nRestoring original name '{}'...", original_name);
    let restore = client
        .rename_symbol(addr, &original_name)
        .unwrap_or_else(|e| {
            eprintln!("Restore failed: {e}");
            std::process::exit(1);
        });
    println!("  renamed={}, name=\"{}\"", restore.renamed, restore.name);

    // 7. Final verification
    let final_check = client.get_symbol(addr).unwrap();
    if let Some(s) = &final_check.symbol {
        println!("  Verified: symbol restored to '{}'", s.name);
    }

    // 8. Show a second page of symbols for pagination demo
    let page2 = client.list_symbols(0, u64::MAX, 10, 20).unwrap();
    println!(
        "\nPagination: symbols 20..30 ({} returned):",
        page2.symbols.len()
    );
    for s in &page2.symbols {
        println!("  0x{:08x}  {}", s.address, s.name);
    }

    println!("\nDone.");
}
